sklearn/linear_model/stochastic_gradient.py

```diff
@@ -211,105 +218,96 @@ def _allocate_parameter_mem(self, n_classes, n_features, coef_init=None,
                                                dtype=np.float64,
                                                order="C")
 
+    def _partial_fit(self, X, y, alpha, C, coef_init, intercept_init, average_coef_init,
+                     average_intercept_init, loss, learning_rate, n_iter,
+                     pos_weight, neg_weight, sample_weight):
+        """Fit a X and y"""
+        n_samples, n_features = X.shape
 
-def _make_dataset(X, y_i, sample_weight):
-    """Create ``Dataset`` abstraction for sparse and dense inputs.
-
-    This also returns the ``intercept_decay`` which is different
-    for sparse datasets.
-    """
-    if sp.issparse(X):
-        dataset = CSRDataset(X.data, X.indptr, X.indices, y_i, sample_weight)
-        intercept_decay = SPARSE_INTERCEPT_DECAY
-    else:
-        dataset = ArrayDataset(X, y_i, sample_weight)
-        intercept_decay = 1.0
-    return dataset, intercept_decay
-
+        # if average is not true, average_coef, and average_intercept will be
+        # unused
+        if not self.average:
+            if coef_init is None:
+                coef_init = np.zeros(n_features, dtype=np.float64, order="C")
+            if intercept_init is None:
+                intercept_init = 0.0
+        else:
+            if coef_init is None:
+                coef_init = np.zeros(n_features, dtype=np.float64, order="C")
+            if intercept_init is None:
+                intercept_init = 0.0
+            if average_coef_init is None:
+                coef_init = np.zeros(n_features, dtype=np.float64, order="C")
+            if intercept_init is None:
```

Comment:
You're absolutely right, I'll change this. Also, now that I look at it, it makes no sense to have the same chunk of code in both logical blocks, so I'll extract that.